### PR TITLE
feat(openapi): implement putting binary file to OSS

### PR DESCRIPTION
**Pull request by graczhual**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2mewvhn 二进制文件上传对接: OSS](https://app.clickup.com/t/2mewvhn)

File: graviti/openapi/policy.py [hunk 1]
```diff
@@ -131,6 +216,61 @@ def get_object(self, key: str, _allow_retry: bool = True) -> UserResponse:
                 return self.get_object(key, False)
             raise error from None
 
+    def put_object(self, file: File, pbar: tqdm, _allow_retry: bool = True) -> None:
```
Comment (linjiX) [inline graviti/openapi/policy.py:219]: should we handle pbar in this method? @Lee-000

File: graviti/openapi/policy.py [hunk 2]
```diff
@@ -131,6 +216,61 @@ def get_object(self, key: str, _allow_retry: bool = True) -> UserResponse:
...
+                do(verb, url, headers=headers, data=fp)
+                pbar.update()
+        except ResponseError as error:
+            code = ElementTree.fromstring(error.response.text)[0].text
+            if _allow_retry and code in _OSS_RETRY_CODE:
+                self._clear_put_policy()
+                self.put_object(file, pbar, False)
+            raise error from None
+
+    def put_objects(
```
Comment (linjiX) [inline graviti/openapi/policy.py:258]: do we need this method in this class?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/openapi/policy.py [hunk 3]
```diff
@@ -131,6 +216,61 @@ def get_object(self, key: str, _allow_retry: bool = True) -> UserResponse:
...
+        policy = self._init_put_policy()
+        date = _from_datetime_to_gmt(datetime.now(timezone.utc))
+        verb = "PUT"
+
+        authorization = self._get_authorization(policy, verb, date, key)
+        mime_type = mimetypes.guess_type(path)[0]
+        headers = {
+            "Authorization": authorization,
+            "x-oss-security-token": policy["SecurityToken"],
+            "Date": date,
```
Comment (Lee-000) [inline graviti/openapi/policy.py:242]: Same logic in get_object, can we reuse them?

File: graviti/openapi/policy.py [hunk 4]
```diff
@@ -131,6 +216,61 @@ def get_object(self, key: str, _allow_retry: bool = True) -> UserResponse:
...
+            "Content-Type": mime_type,
+        }
+        url = f"https://{policy['bucket']}.{policy['endpoint']}/{key}"
+
+        try:
+            with path.open("rb") as fp:
+                do(verb, url, headers=headers, data=fp)
```
Comment (linjiX) [inline graviti/openapi/policy.py:249]: do we need `MultipartEncoder` to upload file?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/openapi/policy.py [hunk 5]
```diff
@@ -131,6 +216,61 @@ def get_object(self, key: str, _allow_retry: bool = True) -> UserResponse:
...
+
+        authorization = self._get_authorization(policy, verb, date, key)
+        mime_type = mimetypes.guess_type(path)[0]
+        headers = {
+            "Authorization": authorization,
+            "x-oss-security-token": policy["SecurityToken"],
+            "Date": date,
+            "Content-Type": mime_type,
+        }
+        url = f"https://{policy['bucket']}.{policy['endpoint']}/{key}"
```
Comment (linjiX) [inline graviti/openapi/policy.py:245]: how does the caller get the object name of the uploaded file?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/openapi/policy.py (hunk 4 above)
Comment (graczhual) [reply graviti/openapi/policy.py:249]: no, here we use putObject method instead of postObject method

**Review by linjiX — APPROVED** [review]

File: graviti/openapi/policy.py [hunk 6]
```diff
@@ -50,75 +53,130 @@ def _request_get_policy(self) -> Dict[str, str]:
...
+        dataset = self._dataset
+        return get_object_policy(  # type: ignore[no-any-return]
+            dataset.access_key,
+            dataset.url,
+            dataset.owner,
+            dataset.name,
+            actions="PUT",
+            is_internal=config.is_internal,
+            expired=_EXPIRED_IN_SECOND,
+        )["policy"]
```
Comment (Lee-000) [inline graviti/openapi/policy.py:66]: This seems identical to `_request_get_policy`, why not define a `_request_policy` method and pass the action to it?

File: graviti/openapi/policy.py [hunk 7]
```diff
@@ -50,75 +53,130 @@ def _request_get_policy(self) -> Dict[str, str]:
...
-        get_hmac = self._get_hmac.copy()
-        get_hmac.update(signature.encode("utf-8"))
-        signature = base64.b64encode(get_hmac.digest()).decode("utf-8")
-        return f"OSS {policy['AccessKeyId']}:{signature}"
+        _hmac = policy["hmac"].copy()
+        _hmac.update(signature.encode("utf-8"))
+        signature = base64.b64encode(_hmac.digest()).decode("utf-8")
+        authorization = f"OSS {policy['AccessKeyId']}:{signature}"
+
+        headers = {
```
Comment (Lee-000) [inline graviti/openapi/policy.py:155]: Return the headers here?

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**